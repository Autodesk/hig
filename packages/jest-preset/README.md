# @hig/jest-preset

Jest preset for HIG packages.

## Usage

_Example `package.json`:_

```json
{
  "jest": {
    "preset": "@hig/jest-preset"
  }
}
```
