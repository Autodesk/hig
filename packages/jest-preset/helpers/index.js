/* eslint-disable global-require */
module.exports = {
  takeSnapshotsOf: require("./takeSnapshotsOf")
};
