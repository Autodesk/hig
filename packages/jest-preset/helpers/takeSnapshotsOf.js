const React = require("react");
const renderer = require("react-test-renderer");

module.exports = function takeSnapshotsOf(Component, cases, rendererOptions) {
  cases.forEach(({ desc, props }) => {
    it(desc, () => {
      const vDom = React.createElement(Component, props);
      const tree = renderer.create(vDom, rendererOptions).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
};
