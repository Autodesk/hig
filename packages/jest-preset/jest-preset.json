{
  "collectCoverageFrom": ["**/src/**/*.{js,jsx}"],
  "coverageDirectory": "<rootDir>/coverage",
  "coveragePathIgnorePatterns": [
    "__gemini__",
    "__stories__",
    "docs",
    "examples/redux",
    "node_modules",
    "packages/jest-preset",
    "packages/vanilla",
    "src/playground"
  ],
  "coverageThreshold": {
    "global": {
      "branches": 50,
      "functions": 50,
      "lines": 50,
      "statements": 50
    }
  },
  "moduleNameMapper": {
    "\\.(css|scss|svg)$": "@hig/jest-preset/fileMock.js"
  },
  "modulePaths": ["node_modules"],
  "moduleFileExtensions": ["js", "jsx", "json"],
  "setupFiles": ["raf/polyfill"],
  "setupTestFrameworkScriptFile": "@hig/jest-preset/setupTests.js",
  "testPathIgnorePatterns": [
    "docs",
    "packages/babel-preset",
    "packages/jest-preset",
    "packages/vanilla"
  ],
  "unmockedModulePathPatterns": ["node_modules"]
}
