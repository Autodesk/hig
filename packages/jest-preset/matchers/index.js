require("./toBeConstants");
require("./toImplementHIGInterfaceOf");
require("./toHavePropTypesAndDocgenInfo");
