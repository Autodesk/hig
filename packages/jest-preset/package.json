{
  "private": true,
  "name": "@hig/jest-preset",
  "version": "0.1.0",
  "dependencies": {
    "babel-core": "^6.26.3",
    "babel-jest": "^22.4.3",
    "enzyme": "^2.9.1",
    "jest-enzyme": "^3.8.3",
    "jest-extended": "^0.8.1",
    "raf": "^3.4.0"
  },
  "eslintConfig": {
    "extends": "@hig"
  }
}
