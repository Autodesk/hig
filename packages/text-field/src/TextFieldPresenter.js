import React, { Component } from "react";
import PropTypes from "prop-types";
import cx from "classnames";
import IconButton from "@hig/icon-button";
import Input from "./presenters/Input";

import "./text-field.scss";

function generatedId() {
  return `text-field-${Math.floor(Math.random() * 100000, 5)}`;
}

export default class TextFieldPresenter extends Component {
  static propTypes = {
    /**
     * Prevents user actions on the field
     */
    disabled: PropTypes.bool,
    /**
     * Error text for the field. Setting this value applies error styling to the entire component.
     */
    errors: PropTypes.string,
    /**
     * Allows managing field's focus via prop. Setting this value overrides the internal focus state management.
     */
    focused: PropTypes.bool,
    /**
     * When true, displays passed error text. When false, displays instructions with error styling.
     */
    hideInstructionsOnErrors: PropTypes.bool,
    /**
     * HTML ID attribute
     */
    id: PropTypes.string,
    /**
     * Icon element that precedes the input.
     */
    icon: PropTypes.node,
    /**
     * Instructional text for the field
     */
    instructions: PropTypes.string,
    /**
     * Text describing what the field represents
     */
    label: PropTypes.string,
    /**
     * Name of the field when submitted with a form
     */
    name: PropTypes.string,
    /**
     * Called when user moves focus from the field
     */
    onBlur: PropTypes.func,
    /**
     * Called after user changes the value of the field
     */
    onChange: PropTypes.func,
    /**
     * Called when user clicks the clear button
     */
    onClearButtonClick: PropTypes.func,
    /**
     * Called when user puts focus onto the field
     */
    onFocus: PropTypes.func,
    /**
     * Called as user changes the value of the field
     */
    onInput: PropTypes.func,
    /**
     * Example of what the user should type into the field
     */
    placeholder: PropTypes.string,
    /**
     * Marks input as read-only
     */
    readOnly: PropTypes.bool,
    /**
     * Text describing why the field is required
     */
    required: PropTypes.string,
    /**
     * When true, causes the clear button to appear
     */
    showClearButton: PropTypes.bool,
    /**
     * Corresponds to the type attribute of an <input>. Relevant for designating a password field, for example.
     */
    type: PropTypes.string,
    /**
     * Value of the field
     */
    value: PropTypes.string
  };

  static defaultProps = {
    id: generatedId(),
    type: "text"
  };

  state = {
    focused: false
  };

  handleFocus = event => {
    this.setState({ focused: true });
    if (this.props.onFocus) this.props.onFocus(event);
  };

  handleBlur = event => {
    this.setState({ focused: false });
    if (this.props.onBlur) this.props.onBlur(event);
  };

  hasClearableInput() {
    return (
      this.props.showClearButton &&
      this.props.value &&
      this.props.value.length > 0
    );
  }

  shouldShowInstructions() {
    if (this.props.instructions) {
      if (this.props.errors) {
        return !this.props.hideInstructionsOnErrors;
      }
      return true;
    }

    return false;
  }

  render() {
    const hasClearableInput = this.hasClearableInput();

    const isFocused =
      this.props.focused !== undefined
        ? this.props.focused
        : this.state.focused;

    const {
      errors,
      focused,
      hideInstructionsOnErrors,
      icon,
      instructions,
      label,
      onClearButtonClick,
      required,
      showClearButton,
      ...inputProps
    } = this.props;

    return (
      <div
        className={cx("hig__text-field-v1", {
          "hig__text-field-v1--required": required,
          "hig__text-field-v1--disabled": inputProps.disabled,
          "hig__text-field-v1--clear-button-visible": hasClearableInput,
          "hig__text-field-v1--with-errors": errors
        })}
      >
        <div
          className={cx("hig__text-field-v1__content", {
            "hig__text-field-v1__content--with-icon": icon
          })}
        >
          <div
            className={cx("hig__text-field-v1__input-wrapper", {
              "hig__text-field-v1__input-wrapper--focused": isFocused,
              "hig__text-field-v1__input-wrapper--with-errors": errors,
              "hig__text-field-v1__input-wrapper--disabled": inputProps.disabled
            })}
          >
            {label && <span className="hig__text-field-v1__label-spacer" />}

            {label && (
              <label
                htmlFor={this.props.id}
                className={cx("hig__text-field-v1__label", {
                  "hig__text-field-v1__label--input-focused": isFocused,
                  "hig__text-field-v1__label--required": required,
                  "hig__text-field-v1__label--with-value": this.props.value
                })}
              >
                {label}
              </label>
            )}

            <div className="hig__text-field-v1__input-row">
              {icon && (
                <label
                  className={cx("hig__text-field-v1__icon", {
                    "hig__text-field-v1__icon--disabled": this.props.disabled
                  })}
                  htmlFor={this.props.id}
                >
                  {icon}
                </label>
              )}

              <Input
                ref={input => {
                  this.input = input;
                }}
                {...inputProps}
                onBlur={this.handleBlur}
                onFocus={this.handleFocus}
              />

              {hasClearableInput && (
                <span className="hig__text-field-v1__clear">
                  <IconButton
                    type="transparent"
                    icon="clear-small"
                    title="Clear field"
                    onClick={onClearButtonClick}
                  />
                </span>
              )}
            </div>
          </div>

          {this.shouldShowInstructions() && (
            <p className="hig__text-field-v1__instructions">{instructions}</p>
          )}

          {errors && <p className="hig__text-field-v1__errors">{errors}</p>}

          {required && (
            <p className="hig__text-field-v1__required-notice">{required}</p>
          )}
        </div>
      </div>
    );
  }
}
