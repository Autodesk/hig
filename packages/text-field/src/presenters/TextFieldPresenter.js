import React, { Component } from "react";
import PropTypes from "prop-types";
import cx from "classnames";
import IconButton, { names, types } from "@hig/icon-button";
import { generateId } from "@hig/utils";
import "@hig/icon-button/build/index.css";

import Input from "./Input";
import "./TextFieldPresenter.scss";

export default class TextFieldPresenter extends Component {
  static propTypes = {
    /**
     * Prevents user actions on the field
     */
    disabled: PropTypes.bool,
    /**
     * Error text for the field. Setting this value applies error styling to the entire component.
     */
    errors: PropTypes.string,
    /**
     * Allows managing field's focus via prop. Setting this value overrides the internal focus state management.
     */
    focused: PropTypes.bool,
    /**
     * When true, displays passed error text. When false, displays instructions with error styling.
     */
    hideInstructionsOnErrors: PropTypes.bool,
    /**
     * HTML ID attribute
     */
    id: PropTypes.string,
    /**
     * Icon element that precedes the input.
     */
    icon: PropTypes.node,
    /**
     * Instructional text for the field
     */
    instructions: PropTypes.string,
    /**
     * Text describing what the field represents
     */
    label: PropTypes.string,
    /**
     * Name of the field when submitted with a form
     */
    name: PropTypes.string,
    /**
     * Called when user moves focus from the field
     */
    onBlur: PropTypes.func,
    /**
     * Called after user changes the value of the field
     */
    onChange: PropTypes.func,
    /**
     * Called when user clicks the clear button
     */
    onClearButtonClick: PropTypes.func,
    /**
     * Called when user puts focus onto the field
     */
    onFocus: PropTypes.func,
    /**
     * Called as user changes the value of the field
     */
    onInput: PropTypes.func,
    /**
     * Example of what the user should type into the field
     */
    placeholder: PropTypes.string,
    /**
     * Marks input as read-only
     */
    readOnly: PropTypes.bool,
    /**
     * Text describing why the field is required
     */
    required: PropTypes.string,
    /**
     * When true, causes the clear button to appear
     */
    showClearButton: PropTypes.bool,
    /**
     * Corresponds to the type attribute of an <input>. Relevant for designating a password field, for example.
     */
    type: PropTypes.string,
    /**
     * Value of the field
     */
    value: PropTypes.string
  };

  static defaultProps = {
    clearButtonTitle: "Clear field",
    type: "text"
  };

  state = {
    focused: false
  };

  id = generateId("text-field");

  handleFocus = event => {
    this.setState({ focused: true });
    if (this.props.onFocus) this.props.onFocus(event);
  };

  handleBlur = event => {
    this.setState({ focused: false });
    if (this.props.onBlur) this.props.onBlur(event);
  };

  hasClearableInput() {
    return (
      this.props.showClearButton &&
      this.props.value &&
      this.props.value.length > 0
    );
  }

  shouldShowInstructions() {
    if (this.props.instructions) {
      if (this.props.errors) {
        return !this.props.hideInstructionsOnErrors;
      }
      return true;
    }

    return false;
  }

  render() {
    const hasClearableInput = this.hasClearableInput();

    const isFocused =
      this.props.focused !== undefined
        ? this.props.focused
        : this.state.focused;

    const {
      id = this.id,
      type,
      errors,
      focused,
      hideInstructionsOnErrors,
      icon,
      instructions,
      label,
      onClearButtonClick,
      required,
      showClearButton,
      clearButtonTitle,
      ...inputProps
    } = this.props;

    const wrapperClasses = cx("hig__text-field-v1", {
      "hig__text-field-v1--button": type === "button",
      "hig__text-field-v1--required": required,
      "hig__text-field-v1--disabled": inputProps.disabled,
      "hig__text-field-v1--clear-button-visible": hasClearableInput,
      "hig__text-field-v1--with-errors": errors
    });

    return (
      <div className={wrapperClasses}>
        <div
          className={cx("hig__text-field-v1__content", {
            "hig__text-field-v1__content--with-icon": icon
          })}
        >
          <div
            className={cx("hig__text-field-v1__input-wrapper", {
              "hig__text-field-v1__input-wrapper--focused": isFocused,
              "hig__text-field-v1__input-wrapper--with-errors": errors,
              "hig__text-field-v1__input-wrapper--disabled": inputProps.disabled
            })}
          >
            {label && <span className="hig__text-field-v1__label-spacer" />}

            {label && (
              <label
                htmlFor={id}
                className={cx("hig__text-field-v1__label", {
                  "hig__text-field-v1__label--input-focused": isFocused,
                  "hig__text-field-v1__label--required": required,
                  "hig__text-field-v1__label--with-value": this.props.value
                })}
              >
                {label}
              </label>
            )}

            <div className="hig__text-field-v1__input-row">
              {icon && (
                <label
                  className={cx("hig__text-field-v1__icon", {
                    "hig__text-field-v1__icon--disabled": this.props.disabled
                  })}
                  htmlFor={id}
                >
                  {icon}
                </label>
              )}

              <Input
                id={id}
                type={type}
                {...inputProps}
                onBlur={this.handleBlur}
                onFocus={this.handleFocus}
              />

              {hasClearableInput && (
                <span className="hig__text-field-v1__clear">
                  <IconButton
                    type={types.TRANSPARENT}
                    icon={names.CLEAR_SMALL}
                    title={clearButtonTitle}
                    onClick={onClearButtonClick}
                  />
                </span>
              )}
            </div>
          </div>

          {this.shouldShowInstructions() && (
            <p className="hig__text-field-v1__instructions">{instructions}</p>
          )}

          {errors && <p className="hig__text-field-v1__errors">{errors}</p>}

          {required && (
            <p className="hig__text-field-v1__required-notice">{required}</p>
          )}
        </div>
      </div>
    );
  }
}
