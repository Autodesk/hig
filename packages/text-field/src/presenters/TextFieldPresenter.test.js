import React from "react";
import { mount } from "enzyme";
import { takeSnapshotsOf } from "@hig/jest-preset/helpers";

import TextFieldPresenter from "./TextFieldPresenter";

describe("TextFieldPresenter", () => {
  describe("label", () => {
    describe("when focusing the input", () => {
      it("adds a relevant class to the label", () => {
        const wrapper = mount(<TextFieldPresenter label="Comments" />);
        const input = wrapper.find("input");

        const label = wrapper.find("label");
        expect(label).not.toHaveClassName(
          "hig__text-field-v1__label--input-focused"
        );
        input.simulate("focus");
        expect(label).toHaveClassName(
          "hig__text-field-v1__label--input-focused"
        );
        input.simulate("blur");
        expect(label).not.toHaveClassName(
          "hig__text-field-v1__label--input-focused"
        );
      });
    });

    it("adds a class when value is present", () => {
      const wrapper = mount(
        <TextFieldPresenter
          label="Comments"
          value="These are my comments"
          onChange={jest.fn()}
        />
      );
      const label = wrapper.find("label");
      expect(label).toHaveClassName("hig__text-field-v1__label--with-value");
    });

    it("adds a class when required", () => {
      const wrapper = mount(
        <TextFieldPresenter label="Comments" required="Comments are required" />
      );
      const label = wrapper.find("label");
      expect(label).toHaveClassName("hig__text-field-v1__label--required");
    });
  });

  takeSnapshotsOf(TextFieldPresenter, [
    {
      desc: "renders without props",
      props: {}
    },
    {
      desc: "renders with all props",
      props: {
        disabled: true,
        errors: "error text",
        focused: true,
        hideInstructionsOnErrors: true,
        id: "my-form-id",
        icon: <span data-hig="icon" />,
        instructions: "instructions",
        label: "label text",
        name: "name text",
        onBlur: function onBlur() {},
        onChange: function onChange() {},
        onClearButtonClick: function onClearButtonClick() {},
        onFocus: function onFocus() {},
        onInput: function onInput() {},
        placeholder: "placeholder text",
        readOnly: true,
        required: "required text",
        showClearButton: true,
        type: "button",
        value: "hello world"
      }
    }
  ]);
});
