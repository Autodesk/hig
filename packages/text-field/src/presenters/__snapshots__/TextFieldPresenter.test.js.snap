// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextFieldPresenter renders with all props 1`] = `
<div
  className="hig__text-field-v1 hig__text-field-v1--button hig__text-field-v1--required hig__text-field-v1--disabled hig__text-field-v1--clear-button-visible hig__text-field-v1--with-errors"
>
  <div
    className="hig__text-field-v1__content hig__text-field-v1__content--with-icon"
  >
    <div
      className="hig__text-field-v1__input-wrapper hig__text-field-v1__input-wrapper--focused hig__text-field-v1__input-wrapper--with-errors hig__text-field-v1__input-wrapper--disabled"
    >
      <span
        className="hig__text-field-v1__label-spacer"
      />
      <label
        className="hig__text-field-v1__label hig__text-field-v1__label--input-focused hig__text-field-v1__label--required hig__text-field-v1__label--with-value"
        htmlFor="my-form-id"
      >
        label text
      </label>
      <div
        className="hig__text-field-v1__input-row"
      >
        <label
          className="hig__text-field-v1__icon hig__text-field-v1__icon--disabled"
          htmlFor="my-form-id"
        >
          <span
            data-hig="icon"
          />
        </label>
        <input
          className="hig__text-field-v1__input"
          disabled={true}
          id="my-form-id"
          name="name text"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onInput={[Function]}
          placeholder="placeholder text"
          readOnly={true}
          type="button"
          value="hello world"
        />
        <span
          className="hig__text-field-v1__clear"
        >
          <button
            className="hig__icon-button hig__icon-button--transparent"
            onBlur={undefined}
            onClick={[Function]}
            onFocus={undefined}
            onMouseEnter={undefined}
            onMouseLeave={undefined}
            title="Clear field"
          >
            <span
              className="hig__icon-button__icon"
            >
              <div
                className="hig__icon hig__icon--24-size"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "<svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"24\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" role=\\"img\\" title=\\"clear-small-24\\"><g fill=\\"#2e4258\\"><path d=\\"M12,22A10,10,0,1,1,22,12,10,10,0,0,1,12,22ZM12,3a9,9,0,1,0,9,9A9,9,0,0,0,12,3Z\\"/><polygon points=\\"15.45 7.47 11.97 10.95 8.51 7.47 7.45 8.53 10.91 12.01 7.45 15.47 8.51 16.53 11.97 13.07 15.42 16.53 16.48 15.47 13.03 12.01 16.51 8.53 15.45 7.47\\"/></g></svg>",
                  }
                }
              />
            </span>
          </button>
        </span>
      </div>
    </div>
    <p
      className="hig__text-field-v1__errors"
    >
      error text
    </p>
    <p
      className="hig__text-field-v1__required-notice"
    >
      required text
    </p>
  </div>
</div>
`;

exports[`TextFieldPresenter renders without props 1`] = `
<div
  className="hig__text-field-v1"
>
  <div
    className="hig__text-field-v1__content"
  >
    <div
      className="hig__text-field-v1__input-wrapper"
    >
      <div
        className="hig__text-field-v1__input-row"
      >
        <input
          className="hig__text-field-v1__input hig__text-field-v1__input--no-value"
          id="text-field-1"
          onBlur={[Function]}
          onFocus={[Function]}
          type="text"
        />
      </div>
    </div>
  </div>
</div>
`;
